Contributes script to install IDAES and COIN solvers on Google Colab

adowling2 opened the pull request:
Contributes script to install IDAES and COIN solvers on Google Colab

## Fixes


## Summary/Motivation:
- Installing IDAES (and solvers via get-extensions) is a little quirky on Colab
- We've used (a prior version of) this script for teaching with Google Colab for over a year with >40 student testers

## Changes proposed in this PR:
- Contributes one file

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

adowling2 commented:
@jsiirola is another good reviewer.

I'd like feedback on:
- How should this get tested?
- Are we okay with the API / level of commenting?
- Are we okay with leaving the `install_CBC`, etc. backup methods? The default usage is just to run `install_idaes()` and then `install_ipopt()`

jsiirola reviewed (changes requested):
In addition to the specific comments below, there are several general changes that should be made:
- fix the identified spelling errors
- remove the logic for direct downloads from `ampl.com`
- I would suggest replacing (all?) uses of `os.system` with `subprocess.run`

adowling2 commented on scripts/colab_helper.py:
```diff
@@ -0,0 +1,218 @@
+            # As a backup, try copying the executable from AMPL
+            if on_colab():
+                print("Installing Ipopt via zip file...")
+                os.system('wget -N -q "https://ampl.com/dl/open/ipopt/ipopt-linux64.zip"')
```
I just removed these.

adowling2 commented on scripts/colab_helper.py:
```diff
@@ -0,0 +1,218 @@
+    '''
+
+    # Check if idaes is available. If not, install it
+    if not package_available("idaes"):
```
Thanks for the great suggestion. I implemented it.

adowling2 commented on scripts/colab_helper.py:
```diff
@@ -0,0 +1,218 @@
+            os.system("ln -s /root/.idaes/bin/cbc cbc")
+            os.system("ln -s /root/.idaes/bin/clp clp")
+            os.system("ln -s /root/.idaes/bin/ipopt_l1 ipopt_l1")
+            os.system("ln -s /root/.idaes/bin/dot_sens dot_sens")
```
Also, I want to make these solvers accessible without `import idaes` for two reasons:
* Have the script print the solver versions, which is a good check and helpful for debugging.
* Reuse this script in workshops/classes/tutorials that introduce Pyomo. In my experience, skipping `import idaes` is easier with brand-new Pyomo users.

adowling2 commented:
@jsiirola This is ready for you to review.

@dangunter You might also find this helpful/interesting. I've been using a script like this to install Pyomo and IDAES in Google Colab for my graduate elective.

adowling2 commented:
@ksbeattie @lbianchi-lbl I have black on the most updated version of this one file on my computer; it said no changes. I also confirmed I do not have any uncommitted or unpushed changes. Yet the black test is still failing.

For everyone: I tested this on Google Colab (hence several commits this morning.) I am not sure how to automate testing in Google Colab or if that is really even needed.

andrewlee94 commented:
@adowling2 `black` is very temperamental - you need to have the correct version of `black` installed as the formatting changes slightly from version to version.

adowling2 commented:
@andrewlee94 Now that I have your attention, do you want to review this quickly?

andrewlee94 reviewed (approved):
This looks OK to me, but I am no expert on this sort of stuff.

jsiirola commented on scripts/colab_helper.py:
```diff
@@ -0,0 +1,142 @@
+
+def _update_path():
+    """Add idaes executables to PATH"""
+    if not re.search(re.escape(":/root/.idaes/bin/"), os.environ["PATH"]):
+        os.environ["PATH"] += ":/root/.idaes/bin/"
```
This is a minor nit, but we probably want to put `idaes/bin` first and not last in the path (the idea being that we would be overriding anything that was previously installed into the system)

adowling2 replied:
I fixed this and tested it works on Colab.

jsiirola commented on scripts/colab_helper.py:
```diff
@@ -0,0 +1,142 @@
+
+def _check_available(executable_name):
+    """Utility to check in an executable is available"""
+    return shutil.which(executable_name) or os.path.isfile(executable_name)
```
(More of an FYI than an action) This is find for collab (where we now it is Linux-based).  However, a more portable approach would be to use something akin to Pyomo's `pyomo.common.fileutils.Executable` (which is what Pyomo uses for locating executables)

jsiirola commented on scripts/colab_helper.py:
```diff
@@ -0,0 +1,142 @@
+        )
+        print("idaes was successfully installed")
+        v = subprocess.run(
+            ["idaes", "--version"], check=True, capture_output=True, text=True
+        )
+        print(v.stdout)
+        print(v.stderr)
```
If you are just going to print out the output anyway, why bother with `capture=True`?

jsiirola replied:
On the terminal, the output comes to the terminal if you don't specify `capture_output`.  I wonder if this is different for the collab environment?  If that's the case, we want to at least be consistent and capture/print for *all* subprocess calls.

adowling2 replied:
I expected that behavior on Colab, but no luck when testing in a cell.

adowling2 replied:
@jsiirola, I like your suggestion. I added a verbose flag to display the captured pip, idaes get-extensions, etc. output

jsiirola reviewed (commented):
This is getting pretty close.

lbianchi-lbl commented on scripts/colab_helper.py:
```diff
@@ -0,0 +1,142 @@
+"https://raw.githubusercontent.com/adowling2/idaes-pse/colab-install-script/scripts/colab_helper.py"
+
+"""
+
```
I'd recommend to add some versioning information, e.g. by adding a line here with `__version__ = "0.23.8"` (or any other date-based versioning schema).

adowling2 replied:
Great suggestion

lbianchi-lbl reviewed (changes requested):
I think having this script here can be valuable for users of IDAES on Google Colab in addition to and beyond the original audience. A few general thoughts:

- I don't think there's a way to test this automatically (without an unreasonable amount of effort)
- Since the script is tested and will continue to be used in the future, I think we can pragmatically rely on this "outside testing" to detect possible breaking changes
- However, since we're adding code to the repository that's not being tested by the CI, before merging, we should:
  -  Identify who the maintainer(s) of this script will be, e.g. by adding an entry to the `CODEOWNERS` file
  - Add some kind of simple versioning information to the script itself (since the script won't have access to either the Git history or the `idaes` version at runtime)

adowling2 commented:
@jsiirola @lbianchi-lbl Thanks for the suggested changes. This should be ready.

Changes:
- Capture all subprocess output. Added verbose flag to toggle on output for pip install, idaes get-extensions, etc.
- Added a version
- Added my name in `CODEOWNERS`

adowling2 commented:
@jsiirola @lbianchi-lbl This is ready for your re-review.
